### PR TITLE
CDC sink support for Postgres

### PR DESCRIPTION
**Pull request by heavycrystal**
Also fixes a couple of minor bugs:

1) The list of unchanged toast columns was not being split properly, so in some cases handling of unchanged toast columns were failing.
2) Meaningful error message when watermark column is not in the list of selected columns in the destination table.

### CODE REVIEW

File: flow/connectors/postgres/postgres.go [hunk 1]
```diff
@@ -304,12 +487,38 @@ func (c *PostgresConnector) GetTableSchema(req *protos.GetTableSchemaInput) (*pr
...
 	req *protos.SetupNormalizedTableInput,
 ) (*protos.SetupNormalizedTableOutput, error) {
-	panic("not implemented")
+	normalizedTableNameComponents, err := parseSchemaTable(req.TableIdentifier)
+	if err != nil {
+		return nil, fmt.Errorf("error while parsing table schema and name: %w", err)
+	}
+	tableAlreadyExists, err := c.tableExists(normalizedTableNameComponents)
+	if err != nil {
+		return nil, fmt.Errorf("error occured while checking if normalized table exists: %w", err)
```
[the remark this answers is not in the document]
Comment (heavycrystal) [reply flow/connectors/postgres/postgres.go:496]: fixed

File: flow/connectors/postgres/client.go [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+			lastCP, syncBatchID, flowJobName)
+		if err != nil {
+			return fmt.Errorf("failed to update flow job status: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (c *PostgresConnector) updateNormalizeMetadata(flowJobName string, normalizeBatchID int64, normalizeRecordsTx pgx.Tx) error {
```
[the remark this answers is not in the document]
Comment (heavycrystal) [reply flow/connectors/postgres/client.go:327]: fixed

File: flow/connectors/postgres/client.go [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+	var version int
+	err := c.pool.QueryRow(c.ctx, "SELECT current_setting('server_version_num')::INTEGER").Scan(&version)
+	if err != nil {
+		return false, fmt.Errorf("failed to get server version: %w", err)
+	}
+
+	return version >= majorVersion, nil
+}
+
+func (c *PostgresConnector) updateSyncMetadata(flowJobName string, lastCP int64, syncBatchID int64, syncRecordsTx pgx.Tx) error {
```
[the remark this answers is not in the document]
Comment (heavycrystal) [reply flow/connectors/postgres/client.go:302]: fixed

File: flow/connectors/postgres/client.go [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+	}
+	return nil
+}
+
+func getRawTableIdentifier(jobName string) string {
+	jobName = regexp.MustCompile("[^a-zA-Z0-9]+").ReplaceAllString(jobName, "_")
+	return fmt.Sprintf("%s_%s", rawTablePrefix, strings.ToLower(jobName))
+}
+
+func generateCreateTableSQLForNormalizedTable(sourceTableIdentifier string, sourceTableSchema *protos.TableSchema) string {
```
[the remark this answers is not in the document]
Comment (heavycrystal) [reply flow/connectors/postgres/client.go:220]: fixed

File: flow/connectors/postgres/postgres_cdc_test.go [hunk 5]
```diff
@@ -162,15 +163,12 @@
...
-		fmt.Sprintf("UPDATE %s SET lz4_b = $1 WHERE id = 2", srcTableName), suite.randBytea(65536))
-	suite.failTestError(err)
-	_, err = mutateRecordsTx.Exec(context.Background(),
-		fmt.Sprintf("UPDATE %s SET n_b = $1 WHERE id = 3", srcTableName), suite.randBytea(65536))
-	suite.failTestError(err)
+	mutateRecordsTx.Exec(context.Background(), "UPDATE %s SET n_t = $1 WHERE id = 1",
+		suite.randString(65536))
+	mutateRecordsTx.Exec(context.Background(), "UPDATE %s SET lz4_t = $1, n_b = $2, lz4_b = $3 WHERE id = 3",
+		suite.randString(65536), suite.randBytea(65536), suite.randBytea(65536))
+	mutateRecordsTx.Exec(context.Background(), "UPDATE %s SET n_t = $1, lz4_t = $2, n_b = $3, lz4_b = $4 WHERE id = 4",
```
[the remark this answers is not in the document]
Comment (heavycrystal) [reply flow/connectors/postgres/postgres_cdc_test.go:170]: fixed

File: flow/connectors/postgres/client.go [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+		lsn_offset BIGINT NOT NULL,sync_batch_id BIGINT NOT NULL,normalize_batch_id BIGINT NOT NULL)`
+	rawTablePrefix          = "_peerdb_raw"
+	createInternalSchemaSQL = "CREATE SCHEMA IF NOT EXISTS %s"
+	createRawTableSQL       = `CREATE TABLE IF NOT EXISTS %s.%s(_peerdb_uid TEXT NOT NULL,
+		_peerdb_timestamp BIGINT NOT NULL,_peerdb_destination_table_name TEXT NOT NULL,_peerdb_data JSONB NOT NULL,
+		_peerdb_record_type INTEGER NOT NULL, _peerdb_match_data JSONB,_peerdb_batch_id INTEGER,
+		_peerdb_unchanged_toast_columns TEXT)`
+
+	getLastOffsetSQL            = "SELECT lsn_offset FROM %s.%s WHERE mirror_job_name=$1"
+	getLastSyncBatchID_SQL      = "SELECT sync_batch_id FROM %s.%s WHERE mirror_job_name=$1"
```
[the remark this answers is not in the document]
Comment (heavycrystal) [reply flow/connectors/postgres/client.go:27]: disabled lint

File: flow/connectors/postgres/postgres.go [hunk 7]
```diff
@@ -138,124 +192,253 @@
...
-			return fmt.Errorf("source tables identifier is invalid: %v", srcTableName)
-		}
-		srcTableNames = append(srcTableNames, srcTableName)
+func (c *PostgresConnector) NormalizeRecords(req *model.NormalizeRecordsRequest) (*model.NormalizeResponse, error) {
+	good, err := c.majorVersionCheck(150000)
+	if err != nil {
+		return nil, err
+	}
+	if !good {
+		return nil, fmt.Errorf("Postgres version is not 15 or higher, required for MERGE")
```
[the remark this answers is not in the document]
Comment (heavycrystal) [reply flow/connectors/postgres/postgres.go:337]: disabled lint

File: flow/connectors/postgres/client.go [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+	getLastNormalizeBatchID_SQL = "SELECT normalize_batch_id FROM %s.%s WHERE mirror_job_name=$1"
```
[the remark this answers is not in the document]
Comment (heavycrystal) [reply flow/connectors/postgres/client.go:28]: disabled lint

**Review by iskakaushik — APPROVED** [review]

**Pull request merged**